src/librustc_hir/hir.rs

```diff
@@ -437,6 +437,29 @@ impl GenericParam<'hir> {
     }
 }
 
+pub trait NextTypeParamName {
+    fn next_type_param_name(&self) -> &'static str;
+}
+
+impl NextTypeParamName for &[GenericParam<'_>] {
+    fn next_type_param_name(&self) -> &'static str {
+        // This is the whitelist of possible parameter names that we might suggest.
+        let possible_names = ["T", "U", "V", "X", "Y", "Z", "A", "B", "C", "D", "E", "F", "G"];
+        let used_names = self
+            .iter()
+            .filter_map(|p| match p.name {
+                ParamName::Plain(ident) => Some(ident.name),
+                _ => None,
+            })
+            .collect::<Vec<_>>();
+
+        possible_names
+            .iter()
+            .find(|n| !used_names.contains(&Symbol::intern(n)))
+            .unwrap_or(&"ParamName")
+    }
+}
```

Comment:
Also I don't think this belongs in `rustc_hir` as it is diagnostics code, and not about AST data structures. You can move it to `rustc_trait_selection` instead.

Reply:
It is also used in `typeck` (where I extracted it from in the first place).

Reply:
`rustc_typeck` has a dependency on `rustc_trait_selection`. I personally don't think this as common as `descr` to warrant adding it here.

Reply:
Moved to `rustc_trait_selection` :)